File: compiler/lib/Opcode.hs
module Opcode where

data Opcode
  = Closure
  | Task0
  | Task1
  | Task2
  | Task3
  | Task4
  | Task5
  | Task6
  | Task7
  | Task8
  | Task9
  | Apply
  | Charge
  | Call
  | Push
  | Return

instance Show Opcode where
  show Closure = "closure"
  show Task0   = "task0"
  show Task1   = "task1"
  show Task2   = "task2"
  show Task3   = "task3"
  show Task4   = "task4"
  show Task5   = "task5"
  show Task6   = "task6"
  show Task7   = "task7"
  show Task8   = "task8"
  show Task9   = "task9"
  show Apply   = "apply"
  show Charge  = "charge"
  show Call    = "call"
  show Push    = "push"
  show Return  = "return"

instruction :: Opcode -> Int -> String
instruction opcode int = padding ++ show opcode ++ " " ++ show int
 where
  padding = replicate offset ' '
  offset  = 2

File: machine/opcode/opcode.go
package opcode

import (
	"log"
)

const (
	/* Callables */
	CLOSURE int32 = iota
	TASK0
	TASK1
	TASK2
	TASK3
	TASK4
	TASK5
	TASK6
	TASK7
	TASK8
	TASK9
	APPLY
	CHARGE
	CALL

	/* Stack */
	PUSH

	RETURN
)

var opcodes = map[string]int32{
	"closure": CLOSURE,
	"task0":   TASK0,
	"task1":   TASK1,
	"task2":   TASK2,
	"task3":   TASK3,
	"task4":   TASK4,
	"task5":   TASK5,
	"task6":   TASK6,
	"task7":   TASK7,
	"task8":   TASK8,
	"task9":   TASK9,
	"apply":   APPLY,
	"charge":  CHARGE,
	"call":    CALL,
	"push":    PUSH,
	"return":  RETURN,
}

func FromString(s string) (opcode int32, exists bool) {
	opcode, exists = opcodes[s]
	return
}

func MustFromString(s string) (opcode int32) {
	if opcode, exists := FromString(s); exists {
		return opcode
	}
	log.Fatalf("opcode %s does not exist", s)
	return
}
